src/app/interaction-model/InvokeResponder.h

```diff
@@ -0,0 +1,192 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/data-model/WrappedStructEncoder.h>
+#include <lib/core/CHIPError.h>
+
+namespace chip {
+namespace app {
+namespace InteractionModel {
+
+/// Aids in writing invoke replies:
+///   - Has access to the raw data encoder for response data content (via `ReplyEncoder`)
+///   - `Complete` handling:
+///      - MUST be called exactly once
+///      - called with an error, makes TLV data written to be discarded (Invokes may only either
+///        succeed or fully fail)
+class InvokeResponder
+{
+public:
+    virtual ~InvokeResponder() = default;
+
+    // Copying not allowed since underlying requirement is that on deletion of this
+    // object, a reply will be sent.
+    InvokeResponder(const InvokeResponder &)             = delete;
+    InvokeResponder & operator=(const InvokeResponder &) = delete;
+
+    /// Flush any pending replies before encoding the current reply.
+    ///
+    /// MAY be called at most once.
+    ///
+    /// This function is intended to provided the ability to retry sending a reply
+    /// if a reply encoding fails due to insufficient buffer.
+    ///
+    /// Call this if `Complete(...)` returns CHIP_ERROR_BUFFER_TOO_SMALL and try
+    /// again. If reply data is needed, the complete ReplyEncoder + Complete
+    /// call chain MUST be re-run.
+    virtual CHIP_ERROR FlushPendingReplies() = 0;
+
+    /// Reply with a data payload.
+    ///
+    /// MUST be called at most once per reply.
+    /// Can be called a 2nd time after a `FlushPendingReplies()` call
+    ///
+    ///   - replyCommandId must correspond with the data encoded in the returned encoder
+    ///   - Complete(CHIP_NO_ERROR) MUST be called to flush the reply
+    virtual DataModel::WrappedStructEncoder & ReplyEncoder(CommandId replyCommandId) = 0;
+
+    /// Signal completing of the reply.
+    ///
+    /// MUST be called exactly once to signal a reply to be sent.
+    /// If this returns CHIP_ERROR_BUFFER_TOO_SMALL, this can be called a 2nd time after
+    /// a FlushPendingReplies.
+    ///
+    /// Argument behavior:
+    ///  - If an ERROR is given (i.e. NOT CHIP_NO_ERROR) an error response will be given to the
+    ///    command.
+    ///  - If CHIP_NO_ERROR is given:
+    ///    - if a ReplyEncoder() was called, a data reply will be sent
+    ///    - if no ReplyEncoder() was called, a Success reply will be sent
```

Comment:
Just to confirm we are on the same page here the StatusIF and Cluster Errors would be in the `CHIP_ERROR error` parameter here. 

If so perhaps noting that in the comments might be helps for future implementer of handling a cluster command. Only reason I mention that is because today there is `AddClusterSpecificSuccess` and `AddStatus` that these callers have been using as of today and converting that may not be intuitive to someone that will be migrating code

Reply:
This was my intent, however looking at StatusIB::InitFromChipError there seems to be no way to say `Status::Success` but with a sdk error code, only Status::Failure seems to be covered.

This is quite unfortunate, however it seems we will need to handle it some way. Will take a look at alternatives.

Reply:
Ended up passing StatusIB for now. It has a clear conversion (even if explicit) from CHIP_ERROR so hopefully code will stay readable.